// @roeybiran/launchbar-env/index.js
"use strict";

/**
 * @property {string} path - the absolute path the current action's .lbaction package
 * @property {string} cachePath - the absolute path the current action's cache folder (in ~/Library/Caches/at.obdev.LaunchBar/Actions/)
 * @property {string} supportPath - the absolute path the current action's support folder (~/Library/Application Support/LaunchBar/Action Support/)
 * @property {boolean} isDebugLogEnabled - True is this debugging is enabled for this action.
 * @property {string} applicationPath - the absolute path to the LaunchBar.app bundle.
 * @property {string} scriptType - the script's type ("default"/"suggestions" etc.).
 * @property {boolean} commandKey - True is this key was held down when the action was invoked.
 * @property {boolean} alternateKey - True is this key was held down when the action was invoked.
 * @property {boolean} shiftKey - True is this key was held down when the action was invoked.
 * @property {boolean} controlKey - True is this key was held down when the action was invoked.
 * @property {boolean} spaceKey - True is this key was held down when the action was invoked.
 * @property {boolean} actionRunsInBackground - True if the action is running in the background.
 * @property {boolean} isLiveFeedbackEnabled - True is Live Feedback is enabled for this action.
 */
module.exports = {
  actionPath: process.env.LB_ACTION_PATH,
  cachePath: process.env.LB_CACHE_PATH,
  supportPath: process.env.LB_SUPPORT_PATH,
  isDebugLogEnabled: process.env.LB_DEBUG_LOG_ENABLED === "1",
  applicationPath: process.env.LB_LAUNCHBAR_PATH,
  scriptType: process.env.LB_SCRIPT_TYPE,
  commandKey: process.env.LB_OPTION_COMMAND_KEY === "1",
  alternateKey: process.env.LB_OPTION_ALTERNATE_KEY === "1",
  shiftKey: process.env.LB_OPTION_SHIFT_KEY === "1",
  controlKey: process.env.LB_OPTION_CONTROL_KEY === "1",
  spaceKey: process.env.LB_OPTION_SPACE_KEY === "1",
  actionRunsInBackground: process.env.LB_OPTION_RUN_IN_BACKGROUND === "1",
  isLiveFeedbackEnabled: process.env.LB_OPTION_LIVE_FEEDBACK === "1"
};
